bolt-modules/boltlib/lib/puppet/functions/apply_prep.rb

```diff
@@ -70,12 +73,47 @@ def apply_prep(target_spec)
 
           unless need_install.empty?
             need_install_targets = need_install.map(&:target)
-            run_task(executor, need_install_targets, 'puppet_agent::install')
-            # Service task works best when targets have puppet-agent feature
+            # lazy-load expensive gem code
+            require 'concurrent'
+            pool = Concurrent::ThreadPoolExecutor.new
+
+            hooks = need_install_targets.map do |t|
+              begin
+                opts = t.plugin_hooks&.fetch('puppet_library')
+                inventory.plugins.get_hook(opts['plugin'], :load_task).call(opts)
+                { 'target' => t,
+                  'hook' => inventory.plugins.get_hook(opts['plugin'], :puppet_library) }
+              rescue Error => e
+                Bolt::Result.from_exception(t, e)
+              end
+            end
+
+            hook_errors, ok_hooks = hooks.partition { |h| h.is_a?(Bolt::Result) }
+
+            futures = ok_hooks.map do |hash|
+              Concurrent::Future.execute(executor: pool) do
+                hash['hook'].call(hash['target'],
+                                  hash['target'].plugin_hooks&.fetch('puppet_library'),
+                                  self)
+              end
+            end
+
+            results = futures.zip(ok_hooks).map do |f, hash|
+              if f.value
```

Comment:
I think you can just use "`.from exception` here.  `f.value || Bolt::Result.from_exception(hash['target'], f.reason)`